app/Controllers/Http/UsersController.ts

```diff
@@ -0,0 +1,54 @@
+import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext';
+import User from 'App/Models/User';
+import CreateUser from 'App/Validators/CreateUserValidator';
+import LoginUser from 'App/Validators/LoginUserValidator';
+
+export default class UsersController {
```

Comment:
Where is this thing registered in any way? Is it based on folders only?
Or some config thing?

Based on that you have string based pointer to this controller and its methods.

Reply:
Adonis uses the string-based path to call required classes into controllers or models and if you follow the path you will find it is pointing to a folder in the app directory or some other directory with some functionality that is needed once imported into a controller or another file. So that is basically its convention.

Reply:
Do you see any problems with that @arnoldkunihira ?

Reply:
Honestly, I don't see any problems with that convention, just like importing express using **import express from 'express';**. They also use string-based imports and also in the tsconfig.json file, the configurations are as below.
![image](https://user-images.githubusercontent.com/19782721/128644404-a702b45c-7172-4088-bfcb-0e9f4536b024.png)

However, I would love to learn the flaws of using string-based pointers.

Reply:
@arnoldkunihira ok, lets say there are three of you working on this project? And a person moves the file. How is this dected. With a solid contract between file import and the existence of the controller method, you might not have a problem.

But with a string relation that is a weak contract and room for lots of errors. Spelling mistake and so on.

But that being said, adonis might handle this smoothly?

Reply:
You are very right with that analogy, however, for me, I installed what you would call an extension in vscode to minimize the room for making lots of errors such as pointing to the actual method in a controller, spelling mistakes.. etc. This extension points to the actual file whose spelling is the same as how it should be as written by the developer. It works in such a way that if something is wrong in the method and path, then it won't underline the route for you. That is how I find out, the path is wrong or spelling
![image](https://user-images.githubusercontent.com/19782721/128646389-315a3b0f-3cb1-49e5-b7f6-7d5e4eeb586e.png)

Reply:
Compare with this whose **show** method is labeled shows and ProductsController is now ProductController, hence it does not underline it and that is how I know there is a problem.
![image](https://user-images.githubusercontent.com/19782721/128646452-b1f789d4-96b1-4f45-b673-6f413f4900f8.png)

Reply:
So In principle, I would encourage the team, to embrace some good practices such as having this extension so that if one of us makes a mistake while developing software, it can be pinpointed and fixed ASAP.

Reply:
@arnoldkunihira okok sounds good. But getting everyone to use an extension is quite narrow. Because some might not even use vscode :P But what happens when you acctually have an error?

Does the code run but fail at runtime/exectuin of the API?

Reply:
Ideally, with an error, it will highlight (with a red underline) to show there exists an actual error, and one needs to fix it ASAP, I believe there are better ways out there to learn when it comes to the handling of errors and I can not wait to learn them.